### PR TITLE
[Refactor] Introduce composable components

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/lightning-flash/tree/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? [not needed for typos/docs]
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-flash/blob/master/CHANGELOG.md)?

## PR review
 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: flash/core/data/process.py [hunk 1]
```diff
@@ -413,15 +400,6 @@ def per_batch_transform_on_device(self, batch: Any) -> Any:
...
-    def available_data_sources(self) -> Sequence[str]:
-        """Get the list of available data source names for use with this
-        :class:`~flash.core.data.process.Preprocess`.
-
-        Returns:
-            The list of data source names.
-        """
-        return list(self._data_sources.keys())
-
     def data_source_of_name(self, data_source_name: str) -> DataSource:
```
Comment (dlangerm) [inline flash/core/data/process.py:403]: This function still needs to go, a Registry should help with that. We'd probably need to add a deprecation notice though.

File: flash/core/data/data_source.py [hunk 2]
```diff
@@ -352,6 +352,65 @@ def generate_dataset(
...
+    def from_fiftyone(self, *args, **kwargs):
+        self._execute(DefaultDataSources.FIFTYONE, *args, **kwargs)
+
+    def from_dataframe(self, *args, **kwargs):
+        self._execute(DefaultDataSources.DATAFRAME, *args, **kwargs)
+
+    def from_lists(self, *args, **kwargs):
+        self._execute(DefaultDataSources.LISTS, *args, **kwargs)
+
+    def from_sentences(self, *args, **kwargs):
```
Comment (dlangerm) [inline flash/core/data/data_source.py:407]: Not a fan of having application-specific code here, a Registry style implementation might work better and reduce code surface. All these functions are basically just calling the same function with a different key, which sounds like a lookup to me rather than hardcoding all of these.

File: flash/image/classification/adapters.py [hunk 3]
```diff
@@ -327,7 +327,7 @@ def _sanetize_batch_size(self, batch_size: int) -> int:
             warning_cache.warn(
                 "When using a meta-learning training_strategy, the batch_size should be set to 1. "
                 "HINT: You can modify the `meta_batch_size` to 100 for example by doing "
-                f"{type(self._task.task)}" + "(training_strategies_kwargs={'meta_batch_size': 100})"
+                "(training_strategies_kwargs={'meta_batch_size': 100})"
```
Comment (dlangerm) [inline flash/image/classification/adapters.py:330]: unrelated?

File: tests/image/classification/test_data.py [hunk 4]
```diff
@@ -611,3 +618,46 @@ def mixup(batch, alpha=1.0):
...
+
+    assert loader._train_ds
+    assert not loader._val_ds
+    assert not loader._test_ds
+    assert loader._predict_ds
+
+    dm = ImageClassificationDataModule.from_components(
+        loader=loader,
+        preprocess=preprocess,
+    )
```
Comment (dlangerm) [inline tests/image/classification/test_data.py:647]: this interface is nice

**Review by dlangerm — COMMENTED** [review]
Still needs work to remove all the hardcoded `from_xxx` stuff, but I like where this is going

File: flash/core/data/data_source.py (hunk 2 above)
Comment (tchaton) [reply flash/core/data/data_source.py:407]: Yes, I know it is not ideal, but this is primarly for IDE resolution, so contributors can follow the function call and understand where the code goes.

Comment (dlangerm) [reply flash/core/data/data_source.py:407]: Yeah that's entirely fair, it just doesn't seem very scalable because now wouldn't we have to implement a custom "from" call both here and in the data source itself? For instance, I have a "from_config_file" method on a derived DataSource class which parses a yaml and loads up my data module.

I'm thinking it might be preferable in the longrun to abstract these away into a more scalable API? I can imaging tons of different "from_*" calls ending up making a lot of sense for different use cases all with different logic.
